src/Forms/Checkout.js

```diff
@@ -0,0 +1,172 @@
+import React from "react";
+import Paper from '@material-ui/core/Paper';
+import Stepper from '@material-ui/core/Stepper';
+import Step from '@material-ui/core/Step';
+import StepLabel from '@material-ui/core/StepLabel';
+import Button from '@material-ui/core/Button';
+import PersonalDetails from "./PersonalDetails";
+import Dates from "./Dates";
+import Payment from "./Payment";
+import SignUpSuccess from "../SignUpSuccess";
+
+class Checkout extends React.Component {
+  state = {
+    activeStep: 0,
+    formControls: {
+      startDate: {
+        value: '2018-07-01',
+        valid: true,
+        touched: false,
+        errorMessage: 'start date is requried'
+      },
+      endDate: {
+        value: '2018-07-05',
+        valid: true,
+        touched: false,
+        errorMessage: 'end date is required'
+      },
+      firstName: {
+        value: '',
+        valid: false,
+        touched: false,
+        errorMessage: 'first name is required'
+      },
+      lastName: {
+        value: '',
+        valid: false,
+        touched: false,
+        errorMessage: 'last name is required'
+      },
+      email: {
+        value: '',
+        valid: false,
+        touched: false,
+        errorMessage: 'email is required'
+      },
+      gender: {
+        value: '',
+        valid: false,
+        touched: false,
+        errorMessage: 'gender is required'
+      }
+    }
+  };
+
+  handleChange = e => {
+    const name = e.target.name;
+    const value = e.target.value;
+
+    const updatedFormControls = { ...this.state.formControls };
+    const updatedElement = { ...this.state.formControls[name] };
+
+    updatedElement.value = value;
+    updatedElement.valid = value !== '';
+    updatedFormControls[name] = updatedElement;
+
+    this.setState({formControls: updatedFormControls});
+  }
+
+  handleBlur = e => {
+    const name = e.target.name;
+
+    const updatedFormControls = { ...this.state.formControls };
+    const updatedElement = { ...this.state.formControls[name] };
+
+    updatedElement.touched = true;
+    updatedFormControls[name] = updatedElement;
+
+    this.setState({formControls: updatedFormControls});
+  }
```

Comment:
Checkout out [immer](https://github.com/immerjs/immer) for easy immutable state changes. With `immer`, you could rewrite this code as follows:

```jsx
this.setState(
  immer(state => {
    state.formControls[name].touched = true;
  })
);
```

Reply:
I feel like I got smarter reading about immur lol. Will definitely have to look into it more